lib/scanner: Recheck file size and modification time after hashing

calmh opened the pull request:
To catch the case where the file changed. Also make sure we never let a
size-vs-blocklist mismatch slip through.

https://forum.syncthing.net/t/since-0-14-lots-of-sync-conflict-files-appearing/7853/21

AudriusButkevicius commented:
feels a bit of hack, but I guess it's cheap

plouj commented:
Would this not keep hashing until the file stops changing faster than the
hashing can finish?

On Mon, Jul 25, 2016 at 5:37 PM, Audrius Butkevicius <
notifications@github.com> wrote:

> feels a bit of hack, but I guess it's cheap
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/syncthing/syncthing/pull/3443#issuecomment-235093587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARPPIg4kRTilDDNaNEHf4ofj0kxTxBfks5qZSyjgaJpZM4JUgNH>
> .
>

AudriusButkevicius commented:
It will just skip the file all together.

plouj commented:
Ah, ok. I guess outbox must be a list of files to hash.

FWIF I think the size != f.Size assertion should be done by unit tests for
HashFile.

AudriusButkevicius commented:
Outbox is what is already hashed.

plouj commented:
Of course when I say "outbox" I mean "inbox" :)

calmh commented:
I moved the check to where we open the file so we can check the fd, and then actually recalculate the size based on the block list after the hashing. Because we've stat:ed at some point long ago before putting the file in the inbox, and that might be out of date. The state just before hashing gets the latest data, and the state of the same fd afterwards ensures that the file didn't change while we were reading it.

I removed the sizehint to HashFile because it doesn't need it any more, and changed the size hint logic in the actual hasher to allow zero as a valid size hint, as that is a valid file size. Otherwise, a file that was zero bytes at the start but grew during hashing would cause an inconsistent result.

Phew... Lets see how this fares when hammered on.

AudriusButkevicius commented:
Good to merge, or are we waiting for more tests by someone?

calmh commented:
No, lets merge. I've ran the test script for a few hours and it looks good. I'll think on how we can add more regular tests for this. I'll release a 0.14.1 with this merged.

calmh commented:
@st-review merge it

st-review commented:
:ok_hand: Merged as 7aaa1dd8a3a2b8e6b9c69b5e147858cefbb4e708. Thanks, @calmh!
